### PR TITLE
add refrence path auto completion

### PR DESCRIPTION
**Pull request by hszhsh**
https://github.com/TypeStrong/atom-typescript/issues/1161

Please have a review if this approach is alright. I'd like to reorganize the code and improve this functionality further.

### CODE REVIEW

File: lib/main/atom/autoCompleteProvider.ts [hunk 1]
```diff
@@ -88,6 +89,53 @@ export class AutocompleteProvider implements Provider {
...
+
+  async getPathSuggestions(
+    prefix: string,
+    location: FileLocationQuery): Promise<SuggestionWithDetails[]> {
+    const client = await this.clientResolver.get(location.file)
+    const projectInfo = await client.executeProjectInfo({
+      file: location.file,
+      needFileNameList: true
+    })
+    const filePaths = projectInfo.body!.fileNames!.filter(p => p != location.file);
```
Comment (guncha) [inline lib/main/atom/autoCompleteProvider.ts]: The optional body for the `project-info` request is there for a reason - Typescript server might not always know what project the current file is in. It's up to you to handle it gracefully.

File: lib/main/atom/autoCompleteProvider.ts [hunk 2]
```diff
@@ -88,6 +89,53 @@ export class AutocompleteProvider implements Provider {
...
+    var files: {
+      name: string;
+      relativePath: string;
+      fullPath: string;
+    }[] = []
+    var sourceDir = path.dirname(location.file);
+    filePaths.forEach(p => {
+      let relativePath = path.relative(sourceDir, p).split('\\').join('/');
```
Comment (guncha) [inline lib/main/atom/autoCompleteProvider.ts]: Is this meant to convert from windows paths to unix? If so, there are specialized functions in `path` module for this.

File: lib/main/atom/autoCompleteProvider.ts [hunk 3]
```diff
@@ -88,6 +89,53 @@ export class AutocompleteProvider implements Provider {
...
+    return files.map(file => {
+      let suggestionText = file.relativePath;
+      let suggestion: Suggestion = {
+        text: suggestionText,
+        replacementPrefix: prefix,
+        rightLabelHTML: '<span>' + file.name + '</span>',
+        type: 'import'
+      }
+
+      return suggestion;
```
Comment (guncha) [inline lib/main/atom/autoCompleteProvider.ts]: This whole function is messy. You're mixing `const` and `var` and mapping over arrays multiple times and creating unnecessary intermediate objects. You should also separate the relative path logic from generating and filtering the suggestions.

**Comment by guncha** [comment]
Also, sorry about taking so long to look at this. I'll try to be better! 😄

**Comment by hszhsh** [comment]
Thanks for you review. I copied most of the code from the old project with little modification, so it's really sprawling. 
For the import path issue, actually lib/util is what I need instead of ../../../../../../lib/util. Because I'll merge multiple paths when my project deployed, and I'd like absolute paths from the path roots instead of relative paths. Is there any other way I can configure it to behave like this?

**Comment by hszhsh** [comment]
Thanks for you review. I copied most of the code from the old project with little modification, so it's really sprawling.
Yes, paths in tsconfig is not handled properly. I think I'll open another project and try to fix this. And I'll try to refactor the code.

**Comment by guncha** [comment]
I think for the time being we can make sure this works for reference comments only and use the default completion for module imports.

P.S. I just reformatted the whole codebase using `prettier` so you're getting some conflicts. I recommend cherry-picking the commit that adds prettier, running it on the codebase and only then merging the master.

**Comment by hszhsh** [comment]
OK, I'll limit this to reference comments and think about module imports further.

File: lib/main/atom/autoCompleteProvider.ts (hunk 2 above)
Comment (hszhsh) [inline lib/main/atom/autoCompleteProvider.ts]: I cannot find such functions in path module, but found makeRelativePath function in main/atom/utils/fs module which does the same thing in the same way.

**Comment by hszhsh** [comment]
Hi, I noticed that you specified ".comment" as disableForSelector for AutocompleteProvider, so it won't work in reference comments. Is it ok to remove the disableForSelector or I should place the code somewhere else?

**Comment by hszhsh** [comment]
And I also find that the code cannot handle /// <reference types="..." /> properly. The issue is the scopes are the same with scopes of /// <reference path="..." />.

**Comment by guncha** [comment]
The `disableForSelector` is just there to cut down on the calls to the `getSuggestions` function. You can safely remove it as long as you make sure the function returns early if it's not a triple slash reference comment.

**Pull request closed without merging**